ibis/pandas/core.py

```diff
@@ -115,311 +116,266 @@
 scalar_types = fixed_width_types + temporal_types
 simple_types = scalar_types + (str, type(None))
 
-_VALID_INPUT_TYPES = (
-    ibis.client.Client,
-    ir.Expr,
-    dt.DataType,
-    type(None),
-    win.Window,
-    tuple,
-) + scalar_types
 
+@functools.singledispatch
+def is_computable_input(arg):
+    """All inputs are not computable without a specific override."""
+    return False
 
-def execute_with_scope(expr, scope, aggcontext=None, clients=None, **kwargs):
-    """Execute an expression `expr`, with data provided in `scope`.
 
-    Parameters
-    ----------
-    expr : ibis.expr.types.Expr
-        The expression to execute.
-    scope : collections.Mapping
-        A dictionary mapping :class:`~ibis.expr.operations.Node` subclass
-        instances to concrete data such as a pandas DataFrame.
-    aggcontext : Optional[ibis.pandas.aggcontext.AggregationContext]
+@is_computable_input.register(ibis.client.Client)
+@is_computable_input.register(ir.Expr)
+@is_computable_input.register(dt.DataType)
+@is_computable_input.register(type(None))
+@is_computable_input.register(win.Window)
+@is_computable_input.register(tuple)
+def is_computable_input_arg(arg):
+    """Return whether `arg` is a valid computable argument."""
+    return True
 
-    Returns
-    -------
-    result : scalar, pd.Series, pd.DataFrame
-    """
-    op = expr.op()
 
-    # Call pre_execute, to allow clients to intercept the expression before
-    # computing anything *and* before associating leaf nodes with data. This
-    # allows clients to provide their own data for each leaf.
-    if clients is None:
-        clients = list(find_backends(expr))
+ibis.util.consume(
```

Comment:
what on earth does this do?

Reply:
oh i see what this is doing, but it's pretty dense. can you add a comment explaining this?

Reply:
For sure.